pg_shard.c

```diff
@@ -394,9 +413,35 @@ ErrorIfQueryNotSupported(Query *queryTree)
 		}
 		else
 		{
-			/* reject subquery, join, function or CTE range table entries */
-			ereport(ERROR, (errmsg("unsupported range table type: %d",
-								   rangeTableEntry->rtekind)));
+			/*
+			 * Error out for rangeTableEntries that we do not support.
+			 * We do not explicitly specify "in FROM clause" in the error detail
+			 * for the features that we do not support at all (SUBQUERY, JOIN).
+			 * We do not need to check for RTE_CTE because all common table expressions
+			 * are rejected above with queryTree->cteList check.
+			 */
+			char *rangeTableEntryErrorDetail = NULL;
+			if (rangeTableEntry->rtekind == RTE_SUBQUERY)
+			{
+				rangeTableEntryErrorDetail = "Subqueries are currently unsupported.";
+			}
+			else if (rangeTableEntry->rtekind == RTE_JOIN)
+			{
+				rangeTableEntryErrorDetail = "Joins are currently unsupported.";
+			}
+			else if (rangeTableEntry->rtekind == RTE_FUNCTION)
+			{
+				rangeTableEntryErrorDetail = "Functions are currently unsupported"
```

Comment:
OK, so this is the odd one out, I guess… How about:

_Functions must not appear in the FROM clause of a distributed query._